File: OpenBurnBarCore/Sources/OpenBurnBarCore/Views/CookingLoader.swift
import SwiftUI

// MARK: - Cooking Loader
//
// A playful, dancing eggs-and-bacon skillet that replaces the mining pickaxe
// loader when the app is in Cooking mode. The whole skillet bounces and sways
// to a gentle spring rhythm — like breakfast dancing in the pan.

public struct CookingLoader: View {
    public typealias Style = MiningPickLoader.Style

    public let style: Style
    public let label: String?
    public let tint: Color?

    public init(_ style: Style = .panel, label: String? = nil, tint: Color? = nil) {
        self.style = style
        self.label = label
        self.tint = tint
    }

    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var bouncePhase: CGFloat = 0

    private var size: CGFloat {
        switch style {
        case .inline: return 24
        case .panel:  return 60
        case .hero:   return 100
        }
    }

    private var labelFont: Font {
        switch style {
        case .inline: return .system(size: 11, weight: .medium, design: .rounded)
        case .panel:  return .system(size: 13, weight: .medium, design: .rounded)
        case .hero:   return .system(size: 16, weight: .semibold, design: .rounded)
        }
    }

    private var spacing: CGFloat {
        switch style {
        case .inline: return 4
        case .panel:  return 10
        case .hero:   return 14
        }
    }

    public var body: some View {
        VStack(spacing: spacing) {
            icon
            if let label, !label.isEmpty {
                Text(label)
                    .font(labelFont)
                    .foregroundStyle(tint ?? .secondary)
                    .multilineTextAlignment(.center)
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(label ?? "Loading")
        .accessibilityAddTraits(.updatesFrequently)
    }

    @ViewBuilder
    private var icon: some View {
        if reduceMotion {
            staticIcon
        } else {
            dancingIcon
        }
    }

    private var staticIcon: some View {
        Image("CookingIconSkillet", bundle: .main)
            .resizable()
            .scaledToFit()
            .frame(width: size, height: size)
            .opacity(0.92)
    }

    private var dancingIcon: some View {
        TimelineView(.animation(minimumInterval: 1.0 / 60, paused: false)) { context in
            let phase = dancingPhase(at: context.date)
            let bounce = bounceOffset(phase)
            let sway = swayRotation(phase)
            let squash = squashScale(phase)

            Image("CookingIconSkillet", bundle: .main)
                .resizable()
                .scaledToFit()
                .frame(width: size, height: size)
                .offset(y: bounce)
                .rotationEffect(.degrees(sway), anchor: .center)
                .scaleEffect(squash, anchor: .center)
        }
        .frame(width: size, height: size)
    }

    // MARK: - Animation curves
    //
    // A single 1.6s loop with overlapping sine waves:
    //   - Bounce: primary vertical hop
    //   - Sway: gentle rotation like a pan rocking on the stove
    //   - Squash: slight stretch on the way up, squash on landing

    private func dancingPhase(at date: Date) -> Double {
        let period: TimeInterval = 1.6
        let t = date.timeIntervalSinceReferenceDate
        return (t.truncatingRemainder(dividingBy: period)) / period
    }

    /// Vertical bounce: egg-like hop, up then settle.
    private func bounceOffset(_ p: Double) -> CGFloat {
        let s = sin(p * .pi * 2)
        // Sharper upward arc, softer landing
        let shaped = s > 0 ? pow(s, 0.7) : -pow(abs(s), 1.4) * 0.3
        return -CGFloat(shaped) * size * 0.12
    }

    /// Gentle rocking sway, offset so it lags the bounce slightly.
    private func swayRotation(_ p: Double) -> Double {
        let lagged = (p + 0.15).truncatingRemainder(dividingBy: 1.0)
        return sin(lagged * .pi * 2) * 5.0
    }

    /// Squash-stretch: taller at apex, squatter at bottom.
    private func squashScale(_ p: Double) -> CGFloat {
        let s = sin(p * .pi * 2)
        // SwiftUI scaleEffect is uniform; we use offset + rotation for character
        return CGFloat(1.0 + s * 0.04)
    }
}

// MARK: - Mode-Aware Loader

public struct ModeAwareLoader: View {
    public let style: MiningPickLoader.Style
    public let label: String?
    public let tint: Color?

    @Environment(\.uiMode) private var uiMode

    public init(_ style: MiningPickLoader.Style = .panel, label: String? = nil, tint: Color? = nil) {
        self.style = style
        self.label = label
        self.tint = tint
    }

    public var body: some View {
        switch uiMode {
        case .cooking:
            CookingLoader(style, label: label, tint: tint)
        case .standard:
            MiningPickLoader(style, label: label, tint: tint)
        }
    }
}

#if DEBUG
#Preview("Cooking loader styles") {
    VStack(spacing: 36) {
        CookingLoader(.inline)
        CookingLoader(.panel, label: "Sizzling…")
        CookingLoader(.hero, label: "Breakfast is almost ready…")
    }
    .padding(40)
}
#endif

File: android/app/src/main/java/com/openburnbar/ui/components/CookingLoader.kt
package com.openburnbar.ui.components

import androidx.compose.animation.core.*
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.openburnbar.R
import com.openburnbar.ui.theme.LocalAuroraReduceMotion
import com.openburnbar.ui.theme.LocalUIMode
import com.openburnbar.ui.theme.UIMode
import kotlin.math.abs
import kotlin.math.sin

enum class CookingLoaderStyle {
    /** Inline spinner (~24dp). Use for buttons, header refreshes, status rows. */
    INLINE,
    /** Page-level loader (~60dp). Use for full-screen loading or panel states. */
    PANEL,
    /** Hero loader (~100dp). Use for prominent first-run or onboarding moments. */
    HERO
}

/**
 * A playful, dancing eggs-and-bacon skillet loading indicator that replaces the standard
 * circular spinner in Cooking mode. The skillet bounces and sways reactively to a gentle rhythm.
 */
@Composable
fun CookingLoader(
    style: CookingLoaderStyle = CookingLoaderStyle.PANEL,
    label: String? = null,
    tint: Color? = null,
    modifier: Modifier = Modifier
) {
    val reduceMotion = LocalAuroraReduceMotion.current

    val size: Dp = when (style) {
        CookingLoaderStyle.INLINE -> 24.dp
        CookingLoaderStyle.PANEL -> 60.dp
        CookingLoaderStyle.HERO -> 100.dp
    }

    val labelTextSize = when (style) {
        CookingLoaderStyle.INLINE -> 11.sp
        CookingLoaderStyle.PANEL -> 13.sp
        CookingLoaderStyle.HERO -> 16.sp
    }

    val labelFontWeight = when (style) {
        CookingLoaderStyle.INLINE, CookingLoaderStyle.PANEL -> FontWeight.Medium
        CookingLoaderStyle.HERO -> FontWeight.SemiBold
    }

    val spacing: Dp = when (style) {
        CookingLoaderStyle.INLINE -> 4.dp
        CookingLoaderStyle.PANEL -> 10.dp
        CookingLoaderStyle.HERO -> 14.dp
    }

    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(spacing)
    ) {
        if (reduceMotion) {
            // Static skillet frame
            Image(
                painter = painterResource(id = R.drawable.ic_cooking_skillet),
                contentDescription = label ?: "Loading",
                modifier = Modifier
                    .size(size)
                    .graphicsLayer { alpha = 0.92f }
            )
        } else {
            // Dancing skillet with physics-based overlaps matching the 1.6s iOS loop
            val infiniteTransition = rememberInfiniteTransition(label = "skillet-dance")
            val animationProgress by infiniteTransition.animateFloat(
                initialValue = 0f,
                targetValue = 1f,
                animationSpec = infiniteRepeatable(
                    animation = tween(1600, easing = LinearEasing),
                    repeatMode = RepeatMode.Restart
                ),
                label = "dance-progress"
            )

            val phase = animationProgress
            val s = sin(phase * Math.PI * 2).toFloat()

            // Vertical bounce: egg-like hop, up then soft landing settle
            val bounceValue = if (s > 0) {
                Math.pow(s.toDouble(), 0.7).toFloat()
            } else {
                -Math.pow(abs(s).toDouble(), 1.4).toFloat() * 0.3f
            }
            val bounceOffset = -bounceValue * size.value * 0.12f

            // Rocking sway: lags the bounce slightly to simulate momentum
            val lagged = (phase + 0.15f) % 1.0f
            val swayRotation = sin(lagged * Math.PI * 2).toFloat() * 5.0f

            // Squash-stretch: taller at apex, squatter at bottom
            val squashScale = 1.0f + s * 0.04f

            Image(
                painter = painterResource(id = R.drawable.ic_cooking_skillet),
                contentDescription = label ?: "Loading",
                modifier = Modifier
                    .size(size)
                    .graphicsLayer {
                        translationY = bounceOffset.dp.toPx()
                        rotationZ = swayRotation
                        scaleX = squashScale
                        scaleY = squashScale
                    }
            )
        }

        if (!label.isNullOrEmpty()) {
            Text(
                text = label,
                fontSize = labelTextSize,
                fontWeight = labelFontWeight,
                color = tint ?: MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.8f),
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(horizontal = 8.dp)
            )
        }
    }
}

/**
 * A mode-aware loading wrapper that inspects the ambient UIMode. Displays the playful
 * dancing CookingLoader when in Cooking mode, falling back to a CircularProgressIndicator.
 */
@Composable
fun ModeAwareLoader(
    modifier: Modifier = Modifier,
    style: CookingLoaderStyle = CookingLoaderStyle.PANEL,
    label: String? = null,
    tint: Color? = null,
    strokeWidth: Dp = 3.dp
) {
    val uiMode = LocalUIMode.current

    if (uiMode == UIMode.COOKING) {
        CookingLoader(
            style = style,
            label = label,
            tint = tint,
            modifier = modifier
        )
    } else {
        Column(
            modifier = modifier,
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            CircularProgressIndicator(
                color = tint ?: MaterialTheme.colorScheme.primary,
                strokeWidth = strokeWidth,
                modifier = Modifier.size(
                    when (style) {
                        CookingLoaderStyle.INLINE -> 20.dp
                        CookingLoaderStyle.PANEL -> 36.dp
                        CookingLoaderStyle.HERO -> 56.dp
                    }
                )
            )
            if (!label.isNullOrEmpty() && style != CookingLoaderStyle.INLINE) {
                Text(
                    text = label,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.Medium,
                    color = tint ?: MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}
